### DISCUSSION TITLE
duckdb

### DISCUSSION THREAD

**Discussion by david-shu** [flet-dev/flet#3881] [Packages] [answered]
duckdb：

https://pypi.org/project/duckdb/

**Comment by ndonkoHenri** [comment, marked as answer]
duckdb 1.5.4 is now available for Flet mobile apps (iOS + Android). Just add it to your dependencies and `flet build` pulls the right iOS/Android [wheels](https://pypi.flet.dev/duckdb) automatically:

```toml
# pyproject.toml
dependencies = [
    "flet",
    "duckdb",
]
```

Notes:
- DuckDB is an **in-process** OLAP database — no server. Use an in-memory DB (`duckdb.connect(":memory:")`)  or a file in your app's storage. (I tested this on Android and iOS with real SQL queries)
- It's a **heavy** package — the engine `.so` is ~15–18 MB per architecture, so it adds notably to app size. Worth it if you need analytical SQL on-device.
- The core engine has no required dependencies. pandas integration works (pandas is on pypi.flet.dev); Arrow output would also need `pyarrow`, which isn't packaged for mobile [yet](https://github.com/flet-dev/flet/discussions/3275).

```py
import duckdb

con = duckdb.connect(":memory:")
con.execute("CREATE TABLE t(id INTEGER, name VARCHAR)")
con.executemany("INSERT INTO t VALUES (?, ?)", [(1, "apple"), (2, "banana")])
print(con.execute("SELECT id, name FROM t ORDER BY id").fetchall())
con.close()
```
